packages/flutter/lib/src/widgets/async.dart

```diff
@@ -182,6 +184,21 @@ enum ConnectionState {
   done,
 }
 
+/// Container class used in [AsyncSnapshot] in debug mode.
+///
+/// The level of indirection allows [AsyncSnapshot] to maintain a `final`
+/// reference to this object, thus adhering to the contract of an `@immutable`
+/// class (at least technically, although not in spirit).
+///
+/// This class is temporary and only exists to facilitate a soft transition
```

Comment:
I am not sure I totally understand the transition path: If the ultimate goal is to make "data" throw if an error is present, shouldn't we for now just fire an assert if "data" is accessed while an error is present?